services/ui-src/src/utils/autosave/autosave.ts

```diff
@@ -1,50 +1,74 @@
 import { FieldValues, UseFormReturn } from "react-hook-form";
 import { Choice, DropdownChoice, ReportStatus } from "types";
-import { AnyObject } from "yup/lib/types";
 
-const valueChanged = (displayValue: any, databaseValue: any) => {
-  return displayValue !== databaseValue;
-};
+type FieldValue = string | DropdownChoice | Choice[] | null;
 
-export const shouldAutosave = (
-  displayValue: any,
-  databaseValue: any,
-  autosave?: boolean,
-  isStateRep?: boolean,
-  isStateUser?: boolean
-) => {
-  return (
-    valueChanged(displayValue, databaseValue) &&
-    autosave &&
-    (isStateRep || isStateUser)
-  );
-};
+interface FieldInfo {
+  name: string;
+  value?: FieldValue;
+  hydrationValue?: FieldValue;
+  shouldClear?: boolean;
+  defaultValue: "" | null;
+}
 
-export const validateAndSetValue = async (
-  form: UseFormReturn<FieldValues, any>,
-  fieldName: string,
-  fieldValue: string | DropdownChoice | Choice[] | null,
-  defaultValue: any
-) => {
-  // check field data validity
-  const fieldDataIsValid = await form.trigger(fieldName);
-
-  // if valid, use; if not, reset to default
-  return fieldDataIsValid ? fieldValue : defaultValue;
-};
+interface Props {
+  form: UseFormReturn<FieldValues, any>;
+  fields: FieldInfo[];
+  report: {
+    id: string | undefined;
+    updateReport: Function;
+  };
+  user: {
+    userName: string | undefined;
+    state: string | undefined;
+    isAuthorizedUser: boolean;
+  };
+}
+
+export const autosaveFieldData = async ({
```

Comment:
Note: I compiled all the methods into one that handles everything autosave. It's a bit big, but it handles everything, so we have central location to look. My first commit made the baseline for it for TextField, then I realized we would need to be able to save multiple fields at once for ChoiceListFields, so I modified it accordingly.

Reply:
Note: None of the tests are updated, and all of the fields except TextField and ChoiceListField are still using the old methods, so there will be failures unless we move those over. Just a proof of concept right now. 🤷